// workbox-cli/build/lib/questions/ask-root-of-web-app.js
"use strict";
/*
  Copyright 2018 Google LLC

  Use of this source code is governed by an MIT-style
  license that can be found in the LICENSE file or at
  https://opensource.org/licenses/MIT.
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.askRootOfWebApp = void 0;
const assert = require("assert");
const fse = require("fs-extra");
const glob = require("glob");
const inquirer_1 = require("inquirer");
const common_tags_1 = require("common-tags");
const errors_1 = require("../errors");
const constants_1 = require("../constants");
const ROOT_PROMPT = 'Please enter the path to the root of your web app:';
// The keys used for the questions/answers.
const questionRootDirectory = 'globDirectory';
const questionManualInput = 'manualDirectoryInput';
/**
 * @return {Promise<Array<string>>} The subdirectories of the current
 * working directory, with hidden and ignored ones filtered out.
 */
async function getSubdirectories() {
    return await new Promise((resolve, reject) => {
        glob('*/', {
            ignore: constants_1.constants.ignoredDirectories.map((directory) => `${directory}/`),
        }, (error, directories) => {
            if (error) {
                reject(error);
            }
            else {
                resolve(directories);
            }
        });
    });
}
/**
 * @return {Promise<Object>} The answers from inquirer.
 */
async function askQuestion() {
    const subdirectories = await getSubdirectories();
    if (subdirectories.length > 0) {
        const manualEntryChoice = 'Manually enter path';
        return inquirer_1.prompt([{
                name: questionRootDirectory,
                type: 'list',
                message: common_tags_1.oneLine `What is the root of your web app (i.e. which directory do
        you deploy)?`,
                choices: subdirectories.concat([
                    new inquirer_1.Separator(),
                    manualEntryChoice,
                ]),
            }, {
                name: questionManualInput,
                when: (answers) => answers.globDirectory === manualEntryChoice,
                message: ROOT_PROMPT,
            }
        ]);
    }
    return inquirer_1.prompt([{
            name: questionRootDirectory,
            message: ROOT_PROMPT,
            default: '.',
        }]);
}
async function askRootOfWebApp() {
    const { manualDirectoryInput, globDirectory } = await askQuestion();
    try {
        const stat = await fse.stat(manualDirectoryInput || globDirectory);
        assert(stat.isDirectory());
    }
    catch (error) {
        throw new Error(errors_1.errors['glob-directory-invalid']);
    }
    return manualDirectoryInput || globDirectory;
}
exports.askRootOfWebApp = askRootOfWebApp;
